### PR TITLE
Start item tree

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by matklad** [comment]
Looks like this can subsume our `AstIdMap` and `HasChildSource` as well, using a nice ECS approach
https://github.com/rust-analyzer/rust-analyzer/pull/3720/commits/2745a45aa185d192a8eba0ffef3735ab75179598

**Comment by Veetaha** [comment]
Data-driven design)

File: crates/ra_hir_def/src/item_tree.rs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+pub struct Const {
+    /// const _: () = ();
+    pub name: Option<Name>,
+    pub visibility: RawVisibility,
+    pub type_ref: TypeRef,
+    pub body: Option<Idx<Expr>>,
+}
+
+pub struct Static {
+    /// const _: () = ();
```
Comment (bjorn3) [inline crates/ra_hir_def/src/item_tree.rs]: s/const/static

**Comment by matklad** [comment]
One thing I am unsure about is how to add handle children. The first way is

```rust
struct Tree {
  structs: Arena<Struct>,
}
struct Struct {
  fields: Arena<Field>
}
```

The second way is

```rust
struct Tree {
  structs: Arena<Struct>,
  fields: Arena<Field>,
}

struct Struct {
  fields: Range<Idx<Field>>
}
```

The advantage of the first approach is that the id of field is local to struct. Adding a new field does not shift other structs. The second scheme is the opposite. 

However, the second scheme seems to be simpler and more efficient (if we discount potential incrementality benefits). In particular, handling sources would be easier. 9cbae00 implements the first version, but I think I'll switch it to the second one.

**Comment by matklad** [comment]
@flodiebold do you have any thoughts on https://github.com/rust-analyzer/rust-analyzer/pull/3720#issuecomment-605020271 ?

**Comment by flodiebold** [comment]
Hmm I'd have thought the incrementality benefits are kind of important... But I guess it really only makes a difference in the moment where one adds a new field somewhere, so it probably doesn't matter that much. And I like the uniformity of the second approach.

**Comment by matklad** [comment]
>Hmm I'd have thought the incrementality benefits are kind of important... 

Yeah, that's why I've started with a nested approach... OTOH, I think incrementality should not really matter in this case: for crate def map, we'll need projection query anyway, for individual items the recompute should not be too bad.

Also, super-long term I actually want to see only single *global* vector for all entities, and using interning to achieve stable indexes ... somehow.

File: crates/ra_hir_def/src/item_tree.rs [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+    pub path: ModPath,
+    pub alias: Option<ImportAlias>,
+    pub visibility: RawVisibility,
+    pub is_glob: bool,
+    pub is_prelude: bool,
+    pub is_extern_crate: bool,
+    pub is_macro_use: bool,
+}
+
+pub struct Function {
```
Comment (Veetaha) [inline crates/ra_hir_def/src/item_tree.rs]: Shouldn't we story unsafety and asynchrony modifiers for `Function` and unsafety for `Trait`?

File: crates/ra_hir_def/src/item_tree.rs [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+    pub name: Name,
+    pub visibility: RawVisibility,
+    pub items: Vec<ModItem>,
+}
+
+pub struct MacroCall {
+    pub name: Option<Name>,
+    pub path: ModPath,
+    pub export: bool,
+    pub builtin: bool,
```
Comment (Veetaha) [inline crates/ra_hir_def/src/item_tree.rs]: I'd suggest to unify the booleans naming, e.g. in `Import` we use `is_*`, but here we don't

File: crates/ra_hir_def/src/item_tree.rs [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+        Idx::from_raw(RawId::from(self.tree.fields.len() as u32))
+    }
+    fn next_variant_idx(&self) -> Idx<Variant> {
+        Idx::from_raw(RawId::from(self.tree.variants.len() as u32))
+    }
+}
+
+fn desugar_future_path(orig: TypeRef) -> Path {
+    let path = path![std::future::Future];
+    let mut generic_args: Vec<_> = std::iter::repeat(None).take(path.segments.len() - 1).collect();
```
Comment (Veetaha) [inline crates/ra_hir_def/src/item_tree.rs]:
```suggestion
    let mut generic_args = vec![None; path.segments.len() - 1];
```

**Comment by matklad** [comment]
Status update: got sidetracked :(

**Comment by matklad** [comment]
Closing due to inactivity, although this is something we absolutely have to do eventually ;-(

My current plan is to nerd-snipe @jonas-schievink into taking over this :-)

**Pull request closed without merging**